pkg/consolegraphql/agent/agent.go

```diff
@@ -0,0 +1,534 @@
+		bearerToken:           bearerToken,
+		host:                  host,
+		port:                  port,
+		addressSpaceNamespace: addressSpaceNamespace,
+		addressSpace:          addressSpace,
+		infraUuid:             infraUuid,
+		stopchan:              make(chan struct{}),
+		stoppedchan:           make(chan struct{}),
+		developmentMode:       developmentMode,
+		commandDelegates:      make(map[string]commandDelegatePair),
+	}
+}
+
+func (aad *amqpAgentDelegate) Collect(handler EventHandler) error {
+	aad.handler = handler
+
+	go func() {
+		defer close(aad.stoppedchan)
+		defer log.Printf("Agent Collector %s - stopped", aad.infraUuid)
+
+		log.Printf("Agent Collector %s - starting", aad.infraUuid)
+		for {
+			select {
+			case <-aad.stopchan:
+				return
+			default:
+				err := aad.doCollect()
+
+				if err == nil {
+					return
+				} else {
+					backoff := computeBackoff(err)
+					log.Printf("Agent Collector %s - restarting - backoff %s(s), %v", aad.infraUuid, backoff, err)
+					if backoff > 0 {
+						time.Sleep(backoff)
+					}
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func computeBackoff(err error) time.Duration {
+	backoff := 5 * time.Second
+	if _, ok := err.(net.Error); ok {
+		backoff = 30 * time.Second
+	}
+	return backoff
+}
+
+func (aad *amqpAgentDelegate) Shutdown() {
+	log.Printf("Agent Collector %s - Shutting down", aad.infraUuid)
+	close(aad.stopchan)
+	for key, commandDelegate := range aad.commandDelegates {
+		commandDelegate.delegate.Shutdown()
+		delete(aad.commandDelegates, key)
+	}
+	<-aad.stoppedchan
+}
+
+func (aad *amqpAgentDelegate) doCollect() error {
+
+	addr := buildAmqpAddress(aad.host, aad.port)
+	log.Printf("Agent Collector %s - connecting %s", aad.infraUuid, addr)
+
+	client, err := amqp.Dial(addr,
+		amqp.ConnTLSConfig(buildTlsConfig(aad.developmentMode)),
+		amqp.ConnSASLXOAUTH2("unused", aad.bearerToken, amqpOverrideSaslFrameSize),
+		amqp.ConnServerHostname(aad.host),
+		amqp.ConnProperty("product", "console-server"),
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = client.Close()
+	}()
+	log.Printf("Agent Collector %s - connected %s", aad.infraUuid, addr)
+
+	// Open a session
+	session, err := client.NewSession()
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+
+	// Create a receiver
+	receiver, err := session.NewReceiver(
+		amqp.LinkSourceAddress(agentDataAddress),
+	)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
+		_ = receiver.Close(ctx)
+		cancel()
+	}()
+
+	log.Printf("Agent Collector %s - commencing collecting", aad.infraUuid)
+	restart := AgentEvent{
+		AddressSpaceNamespace: aad.addressSpaceNamespace,
+		AddressSpace:          aad.addressSpace,
+		InfraUuid:             aad.infraUuid,
+		Type:                  AgentEventTypeRestart,
+	}
+	err = aad.handler(restart)
+	if err != nil {
+		return err
+	}
+
+	housekeepInterval := 1 * time.Minute // TODO Make configurable
```

Comment:
Just so I understand this correctly,this is the 'status collection interval' right?

Reply:
I'll give it a better name.   The agent command delegates belong to a logged (human) user.  They are retained in a cache in a connected state just in case the user invokes another command.  The housekeeper interval refers to the scheduling of a go routine that closes command delegates that have been unused for more than 5 minutes.

Reply:
Thanks for the clarification

Reply:
It is renamed `commandDelegateExpiryPeriod` and controllable via an environment variable.  When command delegates expire, their AMQP connections are closed.